File: backend/src/threatlens/api/routes/exposure.py
"""Exposure Intelligence route: framework + provider status, and an optional real lookup.

Answers "where is this entity exposed", never "is it malicious" (that remains
Threat Intelligence's question) — a separate framework at every layer. Not
integrated into ``/investigate``: a dedicated, isolated lookup.
"""

from __future__ import annotations

import asyncio
from typing import Annotated

from fastapi import APIRouter, Query

from ...exposure import EXPOSURE_FRAMEWORK_VERSION, ExposureService
from ...exposure import build_default_registry as build_exposure_registry
from ...search import detect
from ..schemas import MAX_QUERY_LENGTH, ExposureFrameworkStatus, ExposureProviderStatusInfo

router = APIRouter()

# Exposure Intelligence: a separate framework answering "where is this
# entity exposed", never "is it malicious" (that remains Threat
# Intelligence's question). Built once — Phase 5.1 registers the first
# concrete provider (Shodan); see
# docs/architecture/PHASE-5.0-EXPOSURE-FRAMEWORK.md and
# docs/architecture/PHASE-5.1-SHODAN-PROVIDER.md.
_exposure_registry = build_exposure_registry()
_exposure_service = ExposureService(_exposure_registry)


@router.get("/api/v1/exposure", response_model=ExposureFrameworkStatus)
async def exposure_framework_status(
    value: Annotated[str | None, Query(max_length=MAX_QUERY_LENGTH)] = None,
) -> ExposureFrameworkStatus:
    """Report Exposure Intelligence Framework + provider status, or run a real lookup.

    With no ``value``, this is a pure status probe: framework version,
    registered-provider count, and each provider's health (Shodan today) —
    never an entity lookup. With ``value``, additionally detects the entity
    and runs it through every routed exposure provider, returning their
    merged ``ExposureSummary``. A disabled or unconfigured provider (e.g.
    ``SHODAN_ENABLED=false`` or no ``SHODAN_API_KEY``) yields a well-formed,
    empty or ``unauthorized`` summary — never an error. Still never
    integrated into ``/investigate``.
    """
    providers = _exposure_registry.providers
    health = await asyncio.gather(*(p.health() for p in providers))
    provider_info = [
        ExposureProviderStatusInfo(
            name=provider.metadata.name,
            display_name=provider.metadata.display_name,
            status=snapshot.status,
            detail=snapshot.detail,
        )
        for provider, snapshot in zip(providers, health, strict=True)
    ]

    summary = None
    if value is not None and value.strip():
        entity = detect(value)
        summary = await _exposure_service.investigate(entity)

    count = len(_exposure_registry)
    return ExposureFrameworkStatus(
        status="ready",
        message="No providers configured" if count == 0 else f"{count} provider(s) registered",
        framework_version=EXPOSURE_FRAMEWORK_VERSION,
        providers_registered=count,
        providers=provider_info,
        summary=summary,
    )

File: frontend/lib/api/exposure.ts
// Exposure Intelligence (Phase 5.1 — Shodan is the first provider).
//
// Answers "where is this entity exposed", never "is it malicious" (that
// remains Threat Intelligence's question — a separate framework). Not
// integrated into /investigate: a dedicated, isolated lookup.

import { get } from "./client";
import type { EntityType } from "./investigation";

export type ExposureProviderHealthStatus =
  | "unknown"
  | "operational"
  | "degraded"
  | "unavailable"
  | "disabled";

export interface ExposureProviderStatusInfo {
  name: string;
  display_name: string;
  status: ExposureProviderHealthStatus;
  detail: string | null;
}

export type ExposureResultStatus =
  | "ok"
  | "not_found"
  | "unsupported"
  | "error"
  | "timeout"
  | "rate_limited"
  | "unauthorized";

export type ExposureCapability =
  | "open_ports"
  | "certificates"
  | "passive_dns"
  | "hosting"
  | "asn"
  | "services"
  | "subdomains"
  | "dns_history"
  | "breaches"
  | "credential_exposure"
  | "pastes"
  | "internet_noise";

export interface ExposureFindingErrorInfo {
  message: string;
  retryable: boolean;
  detail: string | null;
}

export interface ExposureEvidence {
  type: string;
  summary: string;
  value: string | null;
  observed_at: string | null;
  data: Record<string, unknown>;
}

export interface ExposureAsset {
  asset_type: string;
  value: string;
  first_seen: string | null;
  last_seen: string | null;
  attributes: Record<string, unknown>;
}

export interface ExposureReference {
  title: string;
  url: string;
  description: string | null;
}

export interface ExposureFinding {
  provider: string;
  provider_display_name: string | null;
  entity_type: EntityType;
  entity_value: string;
  status: ExposureResultStatus;
  error: ExposureFindingErrorInfo | null;
  category: ExposureCapability | null;
  summary: string;
  evidence: ExposureEvidence[];
  assets: ExposureAsset[];
  references: ExposureReference[];
  fetched_at: string | null;
}

export interface ExposureStatistics {
  providers_queried: number;
  providers_ok: number;
  total_findings: number;
  total_assets: number;
  categories: ExposureCapability[];
}

export interface ExposureMetadata {
  entity_type: EntityType;
  entity_value: string;
  generated_at: string;
  framework_version: string;
}

export interface ExposureSummary {
  entity_type: EntityType;
  entity_value: string;
  findings: ExposureFinding[];
  references: ExposureReference[];
  statistics: ExposureStatistics;
  metadata: ExposureMetadata;
}

export interface ExposureFrameworkStatus {
  status: string;
  message: string;
  framework_version: string;
  providers_registered: number;
  providers: ExposureProviderStatusInfo[];
  summary: ExposureSummary | null;
}

/**
 * Fetch Exposure Intelligence Framework + provider status, optionally running
 * a real lookup when `value` is given (e.g. an IP). A disabled or
 * unconfigured provider still returns 200 with a well-formed empty/failed
 * summary — never an error.
 */
export function exposureFrameworkStatus(
  value?: string,
  signal?: AbortSignal,
): Promise<ExposureFrameworkStatus> {
  const trimmed = value?.trim();
  const query = trimmed ? `?value=${encodeURIComponent(trimmed)}` : "";
  return get<ExposureFrameworkStatus>(`/exposure${query}`, signal);
}
